openml/flows/sklearn_converter.py

```diff
@@ -0,0 +1,488 @@
+from collections import OrderedDict, defaultdict
+import importlib
+import inspect
+import json
+import json.decoder
+import six
+import warnings
+import sys
+
+import numpy as np
+import scipy.stats.distributions
+import sklearn.base
+import sklearn.model_selection
+# Necessary to have signature available in python 2.7
+from sklearn.utils.fixes import signature
+
+from .flow import OpenMLFlow
+
+
+if sys.version_info >= (3, 5):
+    from json.decoder import JSONDecodeError
+else:
+    JSONDecodeError = ValueError
+
+
+"""Convert scikit-learn estimators into an OpenMLFlows and vice versa."""
+
+
+def sklearn_to_flow(o):
+
+    if _is_estimator(o):
+        rval = _serialize_model(o)
+    elif isinstance(o, (list, tuple)):
+        rval = [sklearn_to_flow(element) for element in o]
+        if isinstance(o, tuple):
+            rval = tuple(rval)
+    elif o is None:
+        rval = None
+    elif isinstance(o, six.string_types):
+        rval = o
+    elif isinstance(o, (bool, int, float)):
+        rval = o
+    elif isinstance(o, dict):
+        rval = OrderedDict()
+        for key, value in o.items():
+            if not isinstance(key, six.string_types):
+                raise TypeError('Can only use string as keys, you passed '
+                                'type %s for value %s.' %
+                                (type(key), str(key)))
+            key = sklearn_to_flow(key)
+            value = sklearn_to_flow(value)
+            rval[key] = value
+        rval = rval
+    elif isinstance(o, type):
+        rval = serialize_type(o)
+    elif isinstance(o, scipy.stats.distributions.rv_frozen):
+        rval = serialize_rv_frozen(o)
+    # This only works for user-defined functions (and not even partial).
+    # I think this is exactly we want here as there shouldn't be any
+    # built-in or functool.partials in a pipeline
+    elif inspect.isfunction(o):
+        rval = serialize_function(o)
+    elif _is_cross_validator(o):
+        rval = _serialize_cross_validator(o)
+    else:
+        raise TypeError(o, type(o))
+
+    return rval
+
+
+def _is_estimator(o):
+    return (hasattr(o, 'fit') and hasattr(o, 'get_params') and
+            hasattr(o, 'set_params'))
+
+
+def _is_cross_validator(o):
```

Comment:
That's not necessarily true but we can use it for the moment. People can and do implement their own.